File: backend/app/game_rules/achi_muite_hoi.py
from .base import GameRuleBase
from ..models import Player, GameResult


class AchiMuiteHoi(GameRuleBase):
    """あっちむいてホイルール"""

    def judge(self, player1: Player, player2: Player) -> GameResult:
        """勝敗判定（じゃんけん部分）

        Args:
            player1: プレイヤー1
            player2: プレイヤー2

        Returns:
            GameResult: プレイヤー1視点での勝敗結果
        """
        if player1.hand == player2.hand:
            return "draw"

        # 勝利条件（プレイヤー1が勝つ組み合わせ）
        win_conditions = {
            ("rock", "scissors"),
            ("scissors", "paper"),
            ("paper", "rock"),
        }

        if (player1.hand, player2.hand) in win_conditions:
            return "win"
        else:
            return "lose"

    def judge_direction(self, winner: Player, loser: Player) -> bool:
        """方向判定

        Args:
            winner: じゃんけんの勝者
            loser: じゃんけんの敗者

        Returns:
            bool: 方向が一致したらTrue（勝者の勝利）、不一致ならFalse
        """
        return winner.direction == loser.direction

    def get_rule_name(self) -> str:
        """ルール名を取得"""
        return "Achi Muite Hoi (あっちむいてホイ)"

    def get_description(self) -> str:
        """ルール説明を取得"""
        return "Rock-Paper-Scissors + Direction Game. Winner chooses direction, if loser looks the same way, winner wins!"

File: src/game/rules/achi_muite_hoi.ts
// Achi Muite Hoi 方向判定 (Issue #9)。
//
// あっちむいてホイは「じゃんけんで勝った側 (attacker) が方向を指差し、
// 負けた側 (defender) がそれと違う方向を向こうとする」遊び。
// attacker が指した方向と defender が向いた方向が**一致した場合に**
// attacker の勝ち (= 1 point)。一致しなければ point は動かず、再度じゃんけんへ。
//
// 純粋関数として切り出すことで、シーン (AchiMuiteHoiScene) は描画と進行に専念し、
// 「一致したか」のロジックだけここに集約する。

import type { Direction, RoundResult } from '../types'

/**
 * あっちむいてホイの方向判定。
 * 攻め手 (attacker) が指した方向と、受け手 (defender) が向いた方向が一致なら
 * attacker の勝ち。戻り値は **attacker 視点**。
 *
 * - 一致 → 'win'   (attacker が point を取る)
 * - 不一致 → 'lose' (point は動かず再じゃんけん)
 *
 * draw は構造的に発生しない (2 方向の集合は等価か等価でないかの 2 値)。
 */
export const judgeAchiPhase = (
  attackerDir: Direction,
  defenderDir: Direction
): RoundResult => (attackerDir === defenderDir ? 'win' : 'lose')
